File: tunnel-server/src/main/java/com/tuuzed/tunnel/server/tcp/TcpServer.java
package com.tuuzed.tunnel.server.tcp;

import com.tuuzed.tunnel.server.internal.ServerTunnelSessions;
import io.netty.bootstrap.ServerBootstrap;
import io.netty.channel.Channel;
import io.netty.channel.ChannelOption;
import io.netty.channel.nio.NioEventLoopGroup;
import io.netty.channel.socket.nio.NioServerSocketChannel;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class TcpServer {
    @NotNull
    private final TcpTunnelRegistry registry;
    @NotNull
    private final TcpTunnelStats stats;
    @NotNull
    private final ServerBootstrap serverBootstrap;

    public TcpServer(
        @NotNull final NioEventLoopGroup bossGroup,
        @NotNull final NioEventLoopGroup workerGroup
    ) {
        this.registry = new TcpTunnelRegistry();
        this.stats = new TcpTunnelStats();
        this.serverBootstrap = new ServerBootstrap();
        this.serverBootstrap.group(bossGroup, workerGroup)
            .channel(NioServerSocketChannel.class)
            .childOption(ChannelOption.AUTO_READ, true)
            .childOption(ChannelOption.SO_KEEPALIVE, true)
            .childHandler(new TcpServerChannelInitializer(registry, stats));
    }

    @NotNull
    public TcpTunnelRegistry registry() {
        return registry;
    }

    @NotNull
    public TcpTunnelStats stats() {
        return stats;
    }

    public void startTunnel(@Nullable String addr, int port, @NotNull ServerTunnelSessions tunnelSessions) throws Exception {
        final TcpTunnelDescriptor descriptor = new TcpTunnelDescriptor(addr, port, tunnelSessions);
        descriptor.open(serverBootstrap);
        registry.register(port, tunnelSessions, descriptor);
    }

    public void shutdownTunnel(long tunnelToken) {
        registry.unregister(tunnelToken);
    }

    @Nullable
    public Channel getSessionChannel(long tunnelToken, long sessionToken) {
        return registry.getSessionChannel(tunnelToken, sessionToken);
    }

    public void destroy() {
        registry.destroy();
    }


}

File: tunnel2-server/src/main/kotlin/tunnel2/server/tcp/TcpServer.kt
package tunnel2.server.tcp

import io.netty.bootstrap.ServerBootstrap
import io.netty.channel.ChannelInitializer
import io.netty.channel.ChannelOption
import io.netty.channel.nio.NioEventLoopGroup
import io.netty.channel.socket.SocketChannel
import io.netty.channel.socket.nio.NioServerSocketChannel
import tunnel2.server.internal.ServerSessionChannels

class TcpServer(
    bossGroup: NioEventLoopGroup,
    workerGroup: NioEventLoopGroup
) {
    private val serverBootstrap = ServerBootstrap()

    val registry = TcpRegistry()

    init {
        this.serverBootstrap.group(bossGroup, workerGroup)
            .channel(NioServerSocketChannel::class.java)
            .childOption(ChannelOption.AUTO_READ, true)
            .childOption(ChannelOption.SO_KEEPALIVE, true)
            .childHandler(createChannelInitializer())
    }

    fun startTunnel(addr: String?, port: Int, sessionChannels: ServerSessionChannels) {
        val descriptor = TcpDescriptor(addr, port, sessionChannels)
        descriptor.open(serverBootstrap)
        registry.register(port, sessionChannels, descriptor)
    }

    fun destroy() = registry.destroy()

    private fun createChannelInitializer() = object : ChannelInitializer<SocketChannel>() {
        override fun initChannel(ch: SocketChannel?) {
            ch ?: return
            ch.pipeline().addLast(TcpServerChannelHandler(registry))
        }
    }
}